### ISSUE TITLE
Wave around logic for multiclass racing

### ISSUE THREAD

**Issue by joshjaysalazar** [joshjaysalazar/iRacingSafetyCarGenerator#47]
**Is your feature request related to a problem? Please describe.**
When a safety car is deployed, cars trapped between the overall leader and a class leader can disrupt class battles, particularly in multiclass racing.

**Describe the solution you'd like**
Implement wave around logic to account for class leaders and allow affected cars to regain their correct lap position relative to their class.

**Describe alternatives you've considered**
No alternatives currently considered; this is a necessary improvement for realistic multiclass race management.

**Additional context**
N/A

**Comment by ljtijhuis** [comment]
I was not sure if we had another issue in the tracker to do class sorting altogether, but will tie it to this issue:
* Sorting classes when they are mixed together by sending end of line commands to the slower class
* Commands need to be sent in order from class leader to the car in the back so that order is maintained but they are now behind the faster class
* This needs to happen after wave around have occurred to make sure any cars trapped between SC and overall leader are not caught in front of their class lead

Question: how do we know what class is the slower class?

**Comment by elliotb76** [comment]
> Question: how do we know what class is the slower class? Answer: SDK provides an estimated lap time per class.

Does the SDK not have class info per car? How do other overlays show what class other cars are in?

Logic that I think we need:

1. When FCY comes out and the Safety Car picks up Overall Leader
2. Check for (GT3/LMP2) class cars between (GT3/LMP2) Leader and Safety Car based on relative position (Thanks Lars)
3. Send wave around commands for trapped cars in Step 2

NOTE 1: Step 2 can't be determined by overall standings position. It has to happen by position on track, because the trapped cars will always show as behind the GT3 leader on the standings 

NOTE 2: Check for relative position in Step 2 must happen as soon as possible after FCY. This is because if you wait too long to check positions, any leading cars that make it past the Safety Car will join the pack and be instructed by iRacing to filter to the front of their class. IE, GT3 P1 makes it past the SC picking up the overall leader, so they continue around the track, and iRacing sorts them by position, telling GT3 P1 to take their overall position ahead of GT3 P2. This is the cause for leading cars to gain a lap on the rest of their class.

**Comment by ljtijhuis** [comment]
> Does the SDK not have class info per car? How do other overlays show what class other cars are in?

We can see everyone's class and for the classes we have info on their estimated lap time. That is what we can use to make sure we get the classes in the right order and control who in each class should be waved / how to order them appropriately.

Note that we want to generalize this implementation to be applicable to any combination of classes, so we will have to fetch this kind of information dynamically. Alternatively, we'd have to hard code and maintain this information.

> NOTE 2: Check for relative position in Step 2 must happen as soon as possible after FCY. This is because if you wait too long to check positions, any leading cars that make it past the Safety Car will join the pack and be instructed by iRacing to filter to the front of their class. IE, GT3 P1 makes it past the SC picking up the overall leader, so they continue around the track, and iRacing sorts them by position, telling GT3 P1 to take their overall position ahead of GT3 P2. This is the cause for leading cars to gain a lap on the rest of their class.

I am not entirely sure what you mean by this or if it matters. Speaking off the recent example, if P1-3 GT3s passed the SC and we then check who should get a wave around, P4 and up would all be waved around because as seen from the SC, they are all ahead of their class lead. After the wave around, P1-P3 are ahead of the pack as expected. Are you worried iRacing would reshuffle them beyond that?

**Comment by ljtijhuis** [comment]
> I'm not sure if you've watched the replay of Sebring, but that is incorrect. P1-3 continue around, and **iRacing automatically tells them to pass P4 in line**
> 
> This means that from the SC perspective, P4-last are **not** ahead of their class lead

I'm saying this is what we can implement. We know everyone's position on the track, including the SC. So we can implement the logic: if we look backwards from the SC, anyone that is in front of the class lead, we send a wave around command for.

**Comment by ljtijhuis** [comment]
I think we may not be on the same page because timing of the waves matters a lot. I'm assuming we do it ASAP. SC comes out, overall lead makes it to S/F, send waves right then. Alternatively, you wait until the pack gathers up, then send (at which point iRacing may have thrown out some logic of its own). 

This is what I tried to entice input on in the group chat. What is the ideal timing of things, giving cars enough time to follow instructions, not end up conflicting between procedures (e.g. splitting while waves are still happening), not extending the SC for too long.

**Comment by eddielayton** [comment]
Waves should happen as soon as overall leader crosses S/F to be picked up by the SC. as soon as SC is called, the software should check relative position of all cars vs. their class leader and any car between overall leader and class leader should get a wavearound immediately

**Comment by joshjaysalazar** [comment]
I agree with the idea of immediately sending the wave arounds, but I do want to keep the options to delay them for x number of laps. Even though we may not want to use that in our league, I don't want to drop the option.
